### PR TITLE
Auth: Service account store refactor

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**

- renames database filenames to store name
- moves store interface to manager package
- adds an interface to the ProvideAPI constructor
- refactors tests to use the store when necessary
- adds mocks for the new interface implementations in the tests package
- add a new retrieveServiceAccountService to enable injection into the PermissionsServiceAccountService

**Why do we need this feature?**
We are refactoring the service account service to have a private store interface to be more in line with the backend team's new impl/test strategy.

**Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana/issues/58958

Fixes #

**Special notes for your reviewer**:

we need a follow up refactor for creating the serviceaccountimpl and serviceaccounttest packages, same as we have in the users directory according to the backend-platform team. Created a follow up issue - https://github.com/grafana/grafana/issues/59651

### CODE REVIEW

File: pkg/server/wire.go [hunk 1]
```diff
@@ -300,8 +300,8 @@ var wireBasicSet = wire.NewSet(
 	pluginSettings.ProvideService,
 	wire.Bind(new(pluginsettings.Service), new(*pluginSettings.Service)),
 	alerting.ProvideService,
-	database.ProvideServiceAccountsStore,
-	wire.Bind(new(serviceaccounts.Store), new(*database.ServiceAccountsStoreImpl)),
+	serviceaccountsretriever.ProvideService,
```
Comment (eleijonmarck) [inline pkg/server/wire.go:303]: part of retriever that is injected into the ServiceAccountsPermissionsService

File: pkg/services/serviceaccounts/api/api.go [hunk 2]
```diff
@@ -12,32 +14,52 @@ import (
...
+}
+
 func NewServiceAccountsAPI(
 	cfg *setting.Cfg,
-	service serviceaccounts.Service,
+	service service,
 	accesscontrol accesscontrol.AccessControl,
 	accesscontrolService accesscontrol.Service,
 	routerRegister routing.RouteRegister,
-	store serviceaccounts.Store,
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/api/api.go:40]: this is the reason for this refactoring work

File: pkg/services/serviceaccounts/api/api.go [hunk 3]
```diff
@@ -116,7 +137,8 @@ func (api *ServiceAccountsAPI) CreateServiceAccount(c *models.ReqContext) respon
 		}
 	}
 
-	serviceAccount, err := api.store.CreateServiceAccount(c.Req.Context(), c.OrgID, &cmd)
+	serviceAccount, err := api.service.CreateServiceAccount(c.Req.Context(), c.OrgID, &cmd)
+	fmt.Printf("serviceAccount: %v", serviceAccount)
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/api/api.go:141]:
```suggestion
```

File: pkg/services/serviceaccounts/database/store.go [hunk 4]
```diff
@@ -467,9 +466,6 @@ func (s *ServiceAccountsStoreImpl) CreateServiceAccountFromApikey(ctx context.Co
 		}
 
 		if err := s.assignApiKeyToServiceAccount(sess, key.Id, newSA.ID); err != nil {
-			if err := s.userService.Delete(ctx, &user.DeleteUserCommand{UserID: newSA.ID}); err != nil {
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/database/store.go:470]: since we are in a transaction the idea is that the created user in line [464](https://github.com/grafana/grafana/pull/58961/files#diff-8a7c4a45e4d47c36f84d571baa67813c222fc5cd8d02ed9b34bf4fd3e711eb90L464) will be deleted on the rollback, since we have transaction rolling from 463.

File: pkg/services/serviceaccounts/manager/store.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+migration from apikeys to service accounts:
+HideApiKeyTab is used to hide the api key tab in the UI.
+MigrateApiKeysToServiceAccounts migrates all API keys to service accounts.
+MigrateApiKey migrates a single API key to a service account.
+
+// only used for interal api calls
+RevertApiKey reverts a single service account to an API key.
+*/
+type store interface {
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/manager/store.go:22]: this is the meat of this PR

File: pkg/services/serviceaccounts/tests/common.go [hunk 6]
```diff
@@ -125,7 +164,7 @@ func SetupMockAccesscontrol(t *testing.T,
 
 // this is a way to see
 // that the Mock implements the store interface
-var _ serviceaccounts.Store = new(ServiceAccountsStoreMock)
+/* var _ serviceaccounts.Store = new(ServiceAccountsStoreMock) */
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/tests/common.go:167]:
```suggestion
```

File: pkg/services/serviceaccounts/tests/common.go [hunk 7]
```diff
@@ -125,7 +164,7 @@ func SetupMockAccesscontrol(t *testing.T,
 
 // this is a way to see
 // that the Mock implements the store interface
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/tests/common.go:166]:
```suggestion
```

File: pkg/services/serviceaccounts/tests/common.go [hunk 8]
```diff
@@ -123,9 +162,7 @@ func SetupMockAccesscontrol(t *testing.T,
 	return acmock
 }
 
-// this is a way to see
-// that the Mock implements the store interface
-var _ serviceaccounts.Store = new(ServiceAccountsStoreMock)
+/* var _ serviceaccounts.Store = new(ServiceAccountsStoreMock) */
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/tests/common.go:165]:
```suggestion
```

**Review by mildwonkey — APPROVED** [review]
👍 the structure lgtm and it's great to see the service pattern extended! 🏅😁 

I have a random and very nitpicky comment which you should absolutely feel free to ignore, but I find the store method names a bit too wordy. For example, being the serviceAccountStore, I'd go with `Create` instead of `CreateServiceAccount`; `CreateToken`, etc. This is the pattern we (backend squad) gently suggest as a best practice when we're refactoring services. 

Of course naming is hard and subjective and all that 🤷 so again, this is just a nitpick and completely up to your preference.

**Comment by eleijonmarck** [comment]
> 👍 the structure lgtm and it's great to see the service pattern extended! 🏅😁
> 
> I have a random and very nitpicky comment which you should absolutely feel free to ignore, but I find the store method names a bit too wordy. For example, being the serviceAccountStore, I'd go with `Create` instead of `CreateServiceAccount`; `CreateToken`, etc. This is the pattern we (backend squad) gently suggest as a best practice when we're refactoring services.
> 
> Of course naming is hard and subjective and all that 🤷 so again, this is just a nitpick and completely up to your preference.

I gotcha! Do you mind me leaving that as a follow up to the refactoring work. They are way to verbose for sure!

**Comment by mildwonkey** [comment]
> I gotcha! Do you mind me leaving that as a follow up to the refactoring work. They are way to verbose for sure!

@eleijonmarck heck yeah absolutely!

**Comment by eleijonmarck** [comment]
@kalleep it should be up to speed with current main.
Alot of refactoring was done in this PR - https://github.com/grafana/grafana/pull/59910/files#diff-d8c039ddfc3330df45b59cf5ce3a9af979e33cecc9e3d532af8f83a27548b573

Now I have merged the changes

File: pkg/services/featuremgmt/toggles_gen_test.go [hunk 9]
```diff
@@ -20,7 +20,6 @@ import (
 func TestFeatureToggleFiles(t *testing.T) {
 	legacyNames := map[string]bool{
 		"httpclientprovider_azure_auth": true,
-		"service-accounts":              true,
```
Comment (eleijonmarck) [inline pkg/services/featuremgmt/toggles_gen_test.go:23]: this needed to be removed to make tests run.
@bergquist 

Unsure if this is necessary anymore?

File: pkg/services/serviceaccounts/manager/service_test.go [hunk 10]
```diff
@@ -4,17 +4,146 @@ import (
...
+func (f *SecretsCheckerFake) CheckTokens(ctx context.Context) error {
+	return f.ExpectedError
+}
+
 func TestProvideServiceAccount_DeleteServiceAccount(t *testing.T) {
-	t.Run("should call store function", func(t *testing.T) {
-		storeMock := &tests.ServiceAccountsStoreMock{Calls: tests.Calls{}}
-		svc := ServiceAccountsService{store: storeMock}
-		err := svc.DeleteServiceAccount(context.Background(), 1, 1)
+	storeMock := newServiceAccountStoreFake()
```
Comment (IevaVasiljeva) [inline pkg/services/serviceaccounts/manager/service_test.go:122]: nit: this test seems a little bit odd to me, and I think I don't quite understand what we're aiming to test here.

Firstly, it only tests the mock, which imo is not needed. Secondly, it tests both adding and deletion in one test function, but none of the other SA functionality.

**Review by IevaVasiljeva — APPROVED** [review]
I've added some minor comments, but looks good overall! Kalle can probably comment more on the structure, but I think it looks sensible to me.

File: pkg/services/serviceaccounts/manager/service.go [hunk 11]
```diff
@@ -125,10 +144,62 @@ func (sa *ServiceAccountsService) CreateServiceAccount(ctx context.Context, orgI
...
-	return sa.store.DeleteServiceAccount(ctx, orgID, serviceAccountID)
+func (sa *ServiceAccountsService) RetrieveServiceAccount(ctx context.Context, orgID int64, serviceAccountID int64) (*serviceaccounts.ServiceAccountProfileDTO, error) {
+	return sa.store.RetrieveServiceAccount(ctx, orgID, serviceAccountID)
 }
 
 func (sa *ServiceAccountsService) RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error) {
 	return sa.store.RetrieveServiceAccountIdByName(ctx, orgID, name)
 }
+
+func (sa *ServiceAccountsService) DeleteServiceAccount(ctx context.Context, orgID, serviceAccountID int64) error {
```
Comment (Jguer) [inline pkg/services/serviceaccounts/manager/service.go:155]: I would add checks for these methods in Service to check for `orgID = 0` or `serviceAccountID=0` or `apiKeyID=0` and return an `ErrInvalid... ` to ensure the ABI is not misused

File: pkg/services/serviceaccounts/manager/service.go [hunk 12]
```diff
@@ -125,10 +144,62 @@ func (sa *ServiceAccountsService) CreateServiceAccount(ctx context.Context, orgI
...
+}
+
+func (sa *ServiceAccountsService) SearchOrgServiceAccounts(
+	ctx context.Context,
+	orgID int64,
+	query string,
+	filter serviceaccounts.ServiceAccountFilter,
+	page,
+	limit int,
+	signedInUser *user.SignedInUser,
```
Comment (Jguer) [inline pkg/services/serviceaccounts/manager/service.go:170]: Would convert this to a struct query type. (`SearchOrgServiceAccountsQuery`?)

**Review by Jguer — COMMENTED** [review]
Looks good to me, the only change I'd ask before merging would be the invalid value guards in the service

File: pkg/services/serviceaccounts/manager/service.go (hunk 12 above)
Comment (eleijonmarck) [reply pkg/services/serviceaccounts/manager/service.go]: refactored in commit: a7b139c3a154c14fb17d70083de9a6a22b397982

File: pkg/services/serviceaccounts/manager/service.go (hunk 11 above)
Comment (eleijonmarck) [reply pkg/services/serviceaccounts/manager/service.go:171]: added validation in commit: b8b8b7fc9a1bb990820f415362d2cc3ae835d762

File: pkg/services/serviceaccounts/manager/service_test.go (hunk 10 above)
Comment (eleijonmarck) [reply pkg/services/serviceaccounts/manager/service_test.go:120]: talking to the backend team and kalle we should test the struct of interest and mock away the rest to have unit tests only testing the functionality of the struct implementation.
In this case it is the `ServiceAccountsService{}` that is the real thing we want to test and mock away the rest.

Your second point of it only testing deletion is correct. This did not test for anything else than the deletion as well before this refactoring. I could certainly add testing functionality to this refactoring work. If possible we could add another issue for it instead, this would be preferred from my place as we really want to move the store functionality to private as fast as possible. Both in terms of merge conflicts with this PR and both to not have it exposed.

**Review by Jguer — APPROVED** [review]

File: pkg/services/serviceaccounts/manager/service_test.go (hunk 10 above)
Comment (eleijonmarck) [reply pkg/services/serviceaccounts/manager/service_test.go:120]: adding another issue to refactoring backlog for service accounts to make tests for this service.
https://github.com/grafana/grafana/issues/60240
